### PR TITLE
feat(1893,1884): api-catalog - implemented multi schema with subscription

### PR DESCRIPTION
**Pull request by akhilmhdh**
# Closes

1. 1893
2. 1884

# Explain the feature/fix

1. Implemented multiple schema support in backend
2. Added schema validation
3. Added subscription api change details
4. Implemented encryption in headers
5. Fixed timestamp sorting bug

## Does this PR introduce a breaking change

Yes

## Screenshots

<details>
<summary>View Screenshots</summary>

</details>

### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one feature request or issue (no stragglers)?
- [ ] Does the change have appropriate unit tests?
- [ ] Did tests pass?
- [x] Did you update or add any necessary documentation (README.md, WHY.md, etc.)?
- [x] Was this feature demoed and the design review approved?

### CODE REVIEW

File: packages/api-catalog-service/src/dataloaders/subscriptionStatusLoader.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    async (keys: readonly Key[]) => {
+      const ids = keys.map(({ envId }) => envId);
+      const subscriber = keys?.[0]?.userId;
+
+      const docs = await Subscription.aggregate([
+        { $match: { 'subscriptions.environments': { $in: ids }, subscriber } },
+        { $project: { subscription: '$subscriptions.environments' } },
+        { $unwind: '$subscription' },
+        { $unwind: '$subscription' },
+        // { $unwind: '$subscriptions.environments' },
```
Comment (riginoommen) [inline packages/api-catalog-service/src/dataloaders/subscriptionStatusLoader.ts:21]:
```suggestion
```

File: packages/api-catalog-service/src/config.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  const nsEncryptionKey = process.env.NAMESPACE_ENCRYPTION_KEY;
+  if (nsEncryptionKey) {
+    if (Buffer.byteLength(Buffer.from(nsEncryptionKey, 'hex'), 'utf-8') !== 32) {
+      throw Error('Namespace encryption key must  be 256 bits or 32 characters');
+    }
+  }
+
+  return {
+    port: Number(process.env.PORT || 8080),
+    smtpHOST: process.env.SMTP_HOST || '',
```
Comment (riginoommen) [inline packages/api-catalog-service/src/config.ts:32]:
```suggestion
    smtpHost: process.env.SMTP_HOST || '',
```

File: packages/api-catalog-service/src/datasources/namespaceDB.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+} from './types';
+
+export class NamespaceDB extends MongoDataSource<NamespaceModel, IContext> {
+  async getANamespaceById(id: string) {
+    return this.findOneById(id);
+  }
+
+  async getANamespaceBySlug(slug: string) {
+    const namespace = await this.model.findOne({ slug });
+    JSON.stringify(namespace);
```
Comment (riginoommen) [inline packages/api-catalog-service/src/datasources/namespaceDB.ts:21]: JSON.stringify operation seems to be unused programatically

File: packages/api-catalog-service/src/graph/namespace/typedef.graphql [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  createdOn: DateTime
+  createdBy: ApiUserType
+  updatedOn: DateTime
+  updatedBy: ApiUserType
+}
+
+input CreateNamespaceInput {
+  name: String!
+  description: String!
+  # tags: [String]
```
Comment (riginoommen) [inline packages/api-catalog-service/src/graph/namespace/typedef.graphql:67]: Why tags are commented

File: packages/api-catalog-service/src/graph/namespace/typedef.graphql [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+  owners: [ApiOwnerInput]!
+  schemas: [CreateApiSchemaInput]!
+}
+
+input UpdateNamespaceInput {
+  name: String
+  description: String
+  # tags: [String]
```
Comment (riginoommen) [inline packages/api-catalog-service/src/graph/namespace/typedef.graphql:75]: Why tags are commented if its not used we can remove it

File: packages/api-catalog-service/src/graph/types.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+};
+
+export enum ApiCategory {
+  REST = 'REST',
+  GRAPHQL = 'GRAPHQL',
+}
+
+export enum ApiFlags {
+  INTERNAL = 'INTERNAL',
+  DEPRICATED = 'DEPRICATED',
```
Comment (riginoommen) [inline packages/api-catalog-service/src/graph/types.ts:18]:
```suggestion
  DEPRECATED = 'DEPRECATED',
```

**Review by riginoommen — CHANGES_REQUESTED** [review]

File: packages/api-catalog-service/src/dataloaders/subscriptionStatusLoader.ts (hunk 1 above)
Comment (akhilmhdh) [reply packages/api-catalog-service/src/dataloaders/subscriptionStatusLoader.ts]: removed

File: packages/api-catalog-service/src/config.ts (hunk 2 above)
Comment (akhilmhdh) [reply packages/api-catalog-service/src/config.ts]: changed

File: packages/api-catalog-service/src/datasources/namespaceDB.ts (hunk 3 above)
Comment (akhilmhdh) [reply packages/api-catalog-service/src/datasources/namespaceDB.ts]: removed

File: packages/api-catalog-service/src/graph/namespace/typedef.graphql [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+  isDepreciated: Boolean!
+}
+
+# API Namespaces
+type NamespaceType {
+  id: ID
+  name: String
+  slug: String
+  description: String
+  # tags: [String]
```
[the remark this answers is not in the document]
Comment (akhilmhdh) [reply packages/api-catalog-service/src/graph/namespace/typedef.graphql]: Tags are not added for this one. Something to be added laters

File: packages/api-catalog-service/src/graph/namespace/typedef.graphql (hunk 5 above)
Comment (akhilmhdh) [reply packages/api-catalog-service/src/graph/namespace/typedef.graphql]: Removed

File: packages/api-catalog-service/src/graph/types.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+export type OwnerUserType = {
+  user?: Maybe<ApiUserType>;
+  group?: Maybe<ApiEmailGroup>;
+};
+
+export type ApiOwnerType = OwnerMailingType | OwnerUserType;
+
+export type ApiEnvironmentFlags = {
+  isInternal: Scalars['Boolean'];
+  isDepreciated: Scalars['Boolean'];
```
[the remark this answers is not in the document]
Comment (akhilmhdh) [reply packages/api-catalog-service/src/graph/types.ts]: Changed. everywhere

**Comment by akhilmhdh** [comment]
Build is failing for unknown reason. Here is the build in my machine with CI set to true

**Comment by akhilmhdh** [comment]
Screenshot 2022-05-12 at 11 28 13 PM

File: packages/api-catalog-service/src/app.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+import SharedSchema from './graph/typedef.graphql';
+
+export const setupMongoose = async (mongoURI: string, logger: Logger) => {
+  /* Setup database connection */
+  try {
+    await mongoose.connect(mongoURI);
+    logger.info('Connected to database');
+  } catch (error) {
+    logger.error(error, 'Failed to connect to database');
+    process.exit(0);
```
Comment (deshmukhmayur) [inline packages/api-catalog-service/src/app.ts:35]: Use exit status code `1` on failure.
```suggestion
    process.exit(1);
```

**Review by deshmukhmayur — CHANGES_REQUESTED** [review]

File: packages/api-catalog-service/src/app.ts (hunk 9 above)
Comment (akhilmhdh) [reply packages/api-catalog-service/src/app.ts]: Changed

**Review by riginoommen — APPROVED** [review]

**Review by deshmukhmayur — APPROVED** [review]
LGTM

**Pull request merged**